crates/miden-objects/src/account/component/storage/toml/mod.rs

```diff
@@ -0,0 +1,499 @@
+use alloc::collections::{BTreeMap, BTreeSet};
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+
+use miden_core::{Felt, Word};
+use semver::Version;
+use serde::de::Error as _;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use super::super::{
+    AccountStorageSchema,
+    FeltSchema,
+    MapSlotSchema,
+    StorageSlotSchema,
+    StorageValueName,
+    ValueSlotSchema,
+    WordSchema,
+    WordValue,
+};
+use crate::account::component::storage::type_registry::SCHEMA_TYPE_REGISTRY;
+use crate::account::component::{AccountComponentMetadata, SchemaTypeIdentifier};
+use crate::account::{AccountType, StorageSlotName};
+use crate::errors::AccountComponentTemplateError;
+
+mod init_storage_data;
+mod serde_impls;
+
+#[cfg(test)]
+mod tests;
+
+// ACCOUNT COMPONENT METADATA TOML FROM/TO
+// ================================================================================================
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+struct RawAccountComponentMetadata {
+    name: String,
+    description: String,
+    version: Version,
+    supported_types: BTreeSet<AccountType>,
+    #[serde(rename = "storage")]
+    #[serde(default)]
+    storage: RawStorageSchema,
+}
+
+impl AccountComponentMetadata {
+    /// Deserializes `toml_string` and validates the resulting [AccountComponentMetadata]
+    ///
+    /// # Errors
+    ///
+    /// - If deserialization fails
+    /// - If the schema specifies storage slots with duplicates.
+    /// - If the schema contains invalid slot definitions.
+    pub fn from_toml(toml_string: &str) -> Result<Self, AccountComponentTemplateError> {
+        let raw: RawAccountComponentMetadata = toml::from_str(toml_string)
+            .map_err(AccountComponentTemplateError::TomlDeserializationError)?;
+
+        let RawStorageSchema { slot } = raw.storage;
+        let mut fields = Vec::with_capacity(slot.len());
+
+        for slot in slot {
+            fields.push(slot.try_into_slot_schema()?);
+        }
+
+        let storage_schema = AccountStorageSchema::new(fields)?;
+        Ok(Self::new(
+            raw.name,
+            raw.description,
+            raw.version,
+            raw.supported_types,
+            storage_schema,
+        ))
+    }
+
+    /// Serializes the account component metadata into a TOML string.
+    pub fn to_toml(&self) -> Result<String, AccountComponentTemplateError> {
+        let toml =
+            toml::to_string(self).map_err(AccountComponentTemplateError::TomlSerializationError)?;
+        Ok(toml)
+    }
+}
+
+// ACCOUNT STORAGE SCHEMA SERIALIZATION
+// ================================================================================================
+
+// TODO: should this be `[[storage.slots]]`?
```

Comment:
Yeah, maybe `[[storage.slots]]` is better as this seems to be the canonical approach.